### PR TITLE
refactor: rewrite ruby provider in lua

### PR DESCRIPTION
**Pull request by dundargoc**

### CODE REVIEW

File: runtime/autoload/provider/ruby.vim [hunk 1]
```diff
@@ -1,73 +1,16 @@
...
-
-let s:err = ''
-let s:prog = s:detect()
-let s:plugin_path = expand('<sfile>:p:h') . '/script_host.rb'
+let s:prog = v:lua.require'vim.provider.ruby'.detect()
 let g:loaded_ruby_provider = empty(s:prog) ? 1 : 2
-
-if g:loaded_ruby_provider != 2
-  let s:err = 'Cannot find the neovim RubyGem. Try :checkhealth'
-endif
```
Comment (wookayin) [inline runtime/autoload/provider/ruby.vim:70]: Is this part no longer necessary?

File: runtime/lua/vim/provider/ruby.lua [hunk 2]
```diff
@@ -0,0 +1,61 @@
+local M = {}
+local s_err ---@type string?
+local s_host ---@type string?
```
Comment (wookayin) [inline runtime/lua/vim/provider/ruby.lua:3]: technically this is `integer`. (I also missed this too for `perl` and `python` providers)

File: runtime/autoload/provider/ruby.vim (hunk 1 above)
Comment (dundargoc) [reply runtime/autoload/provider/ruby.vim:70]: It wasn't in the python provider. Thought it'd be more consistent if they all followed the same structure.

File: runtime/lua/vim/provider/ruby.lua [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+function M.call(method, args)
+  if s_err then
+    return
+  end
+
+  if not s_host then
+    local ok, result = pcall(vim.fn['remote#host#Require'], 'legacy-ruby-provider') ---@type any, any
+    if not ok then
+      s_err = result
+      vim.api.nvim_echo({ result, 'WarningMsg' }, true, {})
```
Comment (wookayin) [inline runtime/lua/vim/provider/ruby.lua:27]:
```suggestion
      vim.api.nvim_echo({ { result, 'WarningMsg' } }, true, {})
```

Comment (wookayin) [reply runtime/lua/vim/provider/ruby.lua:27]: missed in other providers too..

**Review by wookayin — APPROVED** [review]

File: runtime/lua/vim/provider/ruby.lua (hunk 3 above)
Comment (dundargoc) [reply runtime/lua/vim/provider/ruby.lua]: Fixed all of them.

**Pull request merged**